src/tensors/abstracttensor.jl

```diff
@@ -494,54 +539,39 @@ specialize this method.
 """ Base.similar(::AbstractTensorMap, args...)
 
 function Base.similar(
-        t::AbstractTensorMap, ::Type{T}, codomain::TensorSpace{S}, domain::TensorSpace{S}
-    ) where {T, S}
+        t::AbstractTensorMap, ::Type{T}, codomain::TensorSpace, domain::TensorSpace
+    ) where {T}
     return similar(t, T, codomain ← domain)
 end
+
 # 3 arguments
-function Base.similar(
-        t::AbstractTensorMap, codomain::TensorSpace{S}, domain::TensorSpace{S}
-    ) where {S}
-    return similar(t, similarstoragetype(t), codomain ← domain)
-end
-function Base.similar(t::AbstractTensorMap, ::Type{T}, codomain::TensorSpace) where {T}
-    return similar(t, T, codomain ← one(codomain))
-end
+Base.similar(t::AbstractTensorMap, codomain::TensorSpace, domain::TensorSpace) =
+    similar(t, similarstoragetype(t, scalartype(t)), codomain ← domain)
```

Comment:
Is there a reason this one has `similarstoragetype(t, scalartype(t))` as second argument (which would be equivalent to the shorter `similarstoragetype(t)`), but then the definitions below simply use `scalartype(t)` as second argument?

Reply:
Yes, which might actually be a good reason to revert my last several commits:

For the constructors, we want to take the type of the provided raw vector, so there I need to have `similarstoragetype(data::DenseVector) = typeof(data)`. On the other hand, for the `similar` calls I need to take the type of what would be the result of `similar`, which doesn't always align. This is what was failing - the `PtrArray` data from the manual allocator in TensorOperations has `PtrArray{T,1} <: DenseVector`, but `similar(::PtrArray{T,1})::Vector{T}`.
In order to still force these two methods together, I therefore distinguish by the one and two-argument versions.